pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -568,31 +588,27 @@ func (h *handler) run(infra *infraObject, create bool) (rkev1.RKEMachineStatus,
 			h.dynamic.EnqueueAfter(infra.obj.GetObjectKind().GroupVersionKind(), infra.meta.GetNamespace(), infra.meta.GetName(), 2*time.Second)
 	}
 
-	failedCreate := infra.data.String("status", "failureReason") == string(capierrors.CreateMachineError)
+	failureReasonType := capierrors.CreateMachineError
+	if !create {
+		failureReasonType = capierrors.DeleteMachineError
+	}
+
+	failure := infra.data.String("status", "failureReason") == string(failureReasonType)
 
-	if err := h.apply.WithOwner(infra.obj).ApplyObjects(objects((args.String("providerID") != "" || failedCreate) && create, dArgs)...); err != nil {
-		return rkev1.RKEMachineStatus{}, failedCreate, err
+	if err := h.apply.WithOwner(infra.obj).ApplyObjects(objects((args.String("providerID") != "" || failure) && create, dArgs)...); err != nil {
+		return rkev1.RKEMachineStatus{}, failure, err
 	}
 
-	return dArgs.RKEMachineStatus, failedCreate, err
+	return dArgs.RKEMachineStatus, failure, err
 }
 
-func (h *handler) patchStatus(d data.Object, state rkev1.RKEMachineStatus) error {
+// reconcileStatus will update the infra machine's status by updating each field based on the value of status.
+func reconcileStatus(d data.Object, state rkev1.RKEMachineStatus) error {
 	statusData, err := convert.EncodeToMap(state)
 	if err != nil {
 		return err
 	}
 
-	if state.JobComplete {
```

Comment:
To the reviewer: this was removed because it is only relevant when switching from create to delete, so the logic to reset the `failureMessage` and `failureReason` was moved to the `OnRemove` handler.